### DISCUSSION TITLE
How to enable SPI and I2C on radxa Rock 4?

### DISCUSSION THREAD

**Discussion by AMuellerAtAHS** [MichaIng/DietPi#6229] [Q&A] [answered]
Currently I can't find any explanation, how to use i2c and spi on radxa Rock 4 (SE)? I found some discussions where overlays are used inside /boot/armbianEnv.txt, but I can't find that file. Instead there exists /boot/dietpiEnv.txt. Is it the same? 
Where can I find informations about the parameters? And do I need any extra steps in /boot/dietpi.txt or something else?

**Comment by MichaIng** [comment]
I2C _should_ be enabled by default. For SPI, try adding `spi-spidev` to the `overlays=` line in `/boot/dietpiEnv.txt`. Have a look at available overlays in `/boot/dtb/rockchip/overlay`. Those can be generally added, using the filename minus prefix and ending. But not all of them are compatible with the ROCK 4/RK3399 and even not all of those do actually work: #4596

**Reply by AMuellerAtAHS** [reply]
Thank you very much @MichaIng and sorry for my late response. 
So I tried your suggestions and i2c-7 and spidev1.0 are correctly added to the devices under /dev.

I need 400K baud for i2c and a can interface over mcp2515 and spi. Could you please give me some advice here?

**Reply by MichaIng** [reply]
You do use i2c7? To change the frequency, you'd need to adjust the device tree, or better create an own device tree overlay to enable i2c7 but also change its frequency. Could you check:
```sh
ls -l /proc/device-tree/i2c@ff160000/
cat /proc/device-tree/i2c@ff160000/*frequency*
```

**Reply by Anneress** [reply]
Thanks for your reply, I am on vacation for the next week. 
I will check that when I am back at the office.

**Reply by AMuellerAtAHS** [reply]
Ok here are the outputs:

```
ls -la /proc/device-tree/i2c@ff160000/

total 0
drwxr-xr-x   2 root root  0 Mar 27 14:18  .
drwxr-xr-x 166 root root  0 Mar 27 14:17  ..
-r--r--r--   1 root root  4 Mar 27 14:18 '#address-cells'
-r--r--r--   1 root root  4 Mar 27 14:18  assigned-clock-rates
-r--r--r--   1 root root  8 Mar 27 14:18  assigned-clocks
-r--r--r--   1 root root  4 Mar 27 14:18  clock-frequency
-r--r--r--   1 root root  9 Mar 27 14:18  clock-names
-r--r--r--   1 root root 16 Mar 27 14:18  clocks
-r--r--r--   1 root root 20 Mar 27 14:18  compatible
-r--r--r--   1 root root 16 Mar 27 14:18  interrupts
-r--r--r--   1 root root  4 Mar 27 14:18  name
-r--r--r--   1 root root  4 Mar 27 14:18  phandle
-r--r--r--   1 root root  4 Mar 27 14:18  pinctrl-0
-r--r--r--   1 root root  8 Mar 27 14:18  pinctrl-names
-r--r--r--   1 root root 16 Mar 27 14:18  reg
-r--r--r--   1 root root  4 Mar 27 14:18 '#size-cells'
-r--r--r--   1 root root  5 Mar 27 14:18  status
```
```bash
cat /proc/device-tree/i2c@ff160000/frequency

▒
```

It seems that the output for frequency isn't correct. What should I expect here?

**Reply by MichaIng** [reply]
Ah it is normal that this does not give a readable output, I forgot it.

Hard to know for sure, here the stuff is defined but with variables: https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/arch/arm64/boot/dts/rockchip/rk3399.dtsi?h=linux-5.15.y

Actually it can be simply checked like this:
```sh
stty -F /dev/ttyS7 400000
```
If there is no error messages, 400,000 Hz works.

**Reply by AMuellerAtAHS** [reply]
Thank you for the documentation. 

I tried the command with ttyS7 and i2c-7 both with the same result. Am I missing anything?
```
sudo stty -F /dev/i2c-7 400000

stty: invalid argument ‘400000’
Try 'stty --help' for more information.
```

**Reply by MichaIng** [reply]
I mixed up with UART baudrate. The frequency seems to be 200,000 Hz by default:
```console
# od -bc /proc/device-tree/i2c@ff110000/clock-frequency
0000000 000 003 015 100
         \0 003  \r   @
0000004
```
`003 015 100` (octal) as 32-bit integer: 3*65536 + 13*256 + 64 = 200000

This is for I2C1 since in my case (NanoPi R4S, same SoC) there is no `/proc/device-tree/i2c@ff160000/clock-frequency`, but based on the device tree source and also `/proc/device-tree/i2c@ff160000/assigned-clock-rates` they seem to be identical.

Probably you can verify that the output is the same in your case for `/proc/device-tree/i2c@ff160000/clock-frequency`. If so, I can write a device tree overlay for changing it. But I have not much experience with I2C, whether a doubled frequency just works and does not cause any damage, so all on your own risk 😉.

**Reply by AMuellerAtAHS** [reply]
Hi @MichaIng 

Thank you for your response. I get the following results:
```
# od -bc /proc/device-tree/i2c@ff160000/clock-frequency

0000000 000 006 032 200
         \0 006 032 200
0000004

# od -bc /proc/device-tree/i2c@ff160000/assigned-clock-rates
0000000 013 353 302 000
         \v 353 302  \0
0000004
```

But I don't understand your calculation of the octal value. Can you check and explain it please?

This would be really helpful, if you could help me write a dtoverlay 👍

**Reply by MichaIng** [reply]
Ah that is 400,000 Hz already, so no overlay needed.

> But I don't understand your calculation of the octal value. Can you check and explain it please?

Those are the relevant octal blocks: `006 032 200`

Take one of them:
- Each digit can count up to 7 only, the same way as a digit can count up to 9 only in decimal system. Here we have octal system.
- So a `10` in octal system is an `8` in decimal system. When the first digit reaches `7` and you count one further, the next digit is becoming `1` and the first zero, just like what happens in decimal system when a digit reaches 9.
- An octal `200` hence means `2*8*8=2*8^2=128` in decimal, same as decimal `200` can be split as `2*10*10=2*10^2=200` and in hexadecimal it would be `2*16*16=512`.

Each of the three blocks however are an octet, i.e. consist of 8 bits (or 8 binary digits) only, so it can represent max `2^8=256` numbers, or max `255` (as `0` counts as well, same in IP addresses btw), or `377` in octal. Hence `000 001 000` is `256` as the first block reached `255`, then one further it becomes zero and the next block `1`. Putting all together:
```math
006\ 032\ 200\ (octal\ octet\ blocks)
= 0*8^2+0*8+6\ \ \ \ 0*8^2+3*8+2\ \ \ \ 2*8^2+0*8+0
= 6\ 26\ 128\ (decimal\ octet\ blocks)
= 6*256^2 + 26*256 + 128
= 400000\ (decimal)
```
The fun of interpreting/translating between different positional numbering systems 😄.

**Reply by AMuellerAtAHS** [reply]
Ok thank you really much. I don't understand the calculations for now, but will think about it. Thank you for the explanation. 

So this means, my devicetree overlay with clock-frequency seems to be correct? 

If I use the default devicetree overlay without clock-frequency set, I get the following output:

```
# od -bc /proc/device-tree/i2c@ff160000/clock-frequency
od: /proc/device-tree/i2c@ff160000/clock-frequency: Datei oder Verzeichnis nicht gefunden
```

Does that mean, the default i2c-7 doesn't have clock-frequency? Does that also mean that the clock-frequency isn't rooted down the tree?

**Reply by MichaIng** [reply]
So the file is present when you do apply the `i2c7` overlay and it is missing when you remove it, right? Then all works as expected, that node is disabled by default and the overlay does nothing more than enabling it, so it seems to use 400,000 Hz by default: https://github.com/armbian/build/blob/main/patch/kernel/archive/rockchip64-5.15/general-rockchip-overlays.patch#L225-L241

Or did you do something else than adding `i2c7` to `/boot/dietpiEnv.txt`?

**Reply by AMuellerAtAHS** [reply]
Hmm no. Let me try to explain. 
I already decompiled and changed the original rockchip-i2c7.dtbo and added the clock-frequency parameter.

```
# dtc -I dtb -O dts /boot/dtb-5.15.93-rockchip64/rockchip/overlay/rockchip-i2c7.dtbo -o rockchip-i2c7-400k.dts
# nano rockchip-i2c7-400k.dts

/dts-v1/;

/ {
        compatible = "rockchip,rk3399";

        fragment@0 {
                target-path = "/i2c@ff160000";

                __overlay__ {
                        status = "okay";
                        /* Added this clock-frequency parameter */
                        clock-frequency = <400000>;
                };
        };
};

# armbian-add-overlay rockchip-i2c7-400k.dts 
```
With that user-overlay I get the following:

```
# od -bc /proc/device-tree/i2c@ff160000/clock-frequency

0000000 000 006 032 200
         \0 006 032 200
0000004
```

But with the original rockchip-i2c7.dtbo I get:

```
# od -bc /proc/device-tree/i2c@ff160000/clock-frequency
od: /proc/device-tree/i2c@ff160000/clock-frequency: File or directory not found
```

If the original overlay doesn't have any clock-frequency set. Am I able to set one which is used by the driver?

How can I test the functionality? 

I tried with:
```
# i2cdetect -y 7 
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
00:                         -- -- -- -- -- -- -- --
10: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
20: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
30: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
40: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
50: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
60: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
70: -- -- -- -- -- -- -- --

**Reply by MichaIng** [reply]
Ah okay, so then it is the same as with my NanoPi R4S, which also does not have the `clock-frequency` at this node. I've no idea why and I don't know whether just adding it is actually effective. I suggest you just test it with the MCP2515? I'm also not sure how to test the I2C frequency or do some sorts of dry run from the I2C host itself.

**Reply by AMuellerAtAHS** [reply, marked as answer]
Hi MichaIng, thanks a lot for your help.
Finally my i2c display was broken, when I replaced it, everything works like expected.

**Reply by MichaIng** [reply]
Okay great that it finally works 🙂.
